BJS2-53956: Создание и настройка скиллов

rostex opened the pull request:
Реализованы основные операции с Skill Entity

Добавлены:
- создание Умения;
- получение списка Умений Пользователя;
- получение предложенных Умений;
- приобретение предложенных Умений.

rostex committed:
Реализованы основные операции с Skill Entity

Добавлены:
- операция создания Умения;
- получения списка Умений Пользователя;
- получение предложенных Умений;
- приобретений предложенных Умений.

rostex committed:
Реализованы основные операции с Skill Entity

Добавлены:
- операция создания Умения;
- получения списка Умений Пользователя;
- получение предложенных Умений;
- приобретений предложенных Умений.

rostex committed:
Реализованы основные операции с Skill Entity

Добавлены:
- операция создания Умения;
- получения списка Умений Пользователя;
- получение предложенных Умений;
- приобретений предложенных Умений.

GigaNiga666 commented on src/main/java/school/faang/user_service/controller/SkillController.java:
```diff
@@ -0,0 +1,42 @@
+import java.util.List;
+
+@Component
+@RequiredArgsConstructor
```
Здесь должна быть аннотация RestController и RequestMapping для указания URL пути на этот контроллер

sstani-bgv commented on src/main/java/school/faang/user_service/controller/SkillController.java:
```diff
@@ -0,0 +1,42 @@
+    private final SkillService skillService;
+
+    public SkillDto create(@Valid @RequestBody SkillDto skill) {
+        validateTitle(skill);
```
В Dto всю валидацию надо делать, у тебя там уже NotBlank есть над title. Добавь просто ещё NotNull

rostex replied:
Прочитав документацию и проверив опытным путем (тестами) аннотация NotBlank подразумевает, что поле не null

sstani-bgv replied:
У тебя просто NotBlank проверятся даже не будет => если title придёт как null, то ошибки не будет, это не правильно поведение

rostex replied:
Пруфы есть?

rostex replied:
![NotNull](https://github.com/user-attachments/assets/d82c27ab-bba2-488f-8f99-39a87d316d82)
Опять же в документации указано, что объект не должен быть null и содержать хотябы один не пустой символ
В тестах проверка есть на null и blank. Все работает

sstani-bgv commented on src/main/java/school/faang/user_service/dto/skill/SkillCandidateDto.java:
```diff
@@ -0,0 +1,11 @@
+
+@Data
+@NoArgsConstructor
+public class SkillCandidateDto {
```
Можно сделать record, все аннотации будут работать, проверено)

rostex replied:
Record используется в случае если класс иммутабельный (все поля final, нет сеттеров). DTO подразумеавет под собой, что его поля будут изменяться при переносе данных, соотвественно Record в данном случае не подойдет.

sstani-bgv replied:
На дейлике это обсудим)

sstani-bgv commented on src/main/java/school/faang/user_service/dto/skill/SkillDto.java:
```diff
@@ -0,0 +1,13 @@
+@Data
+@NoArgsConstructor
+public class SkillDto {
+    private Long id;
```
Проверку на null и min добавь

rostex replied:
NotBlank под собой подразумевает, что String не null. По поводу указания MIN сомневаюсь, я думаю достаточно, что оно NotBlank.

sstani-bgv replied:
Не, на поле id добавить надо. Min надо для того, чтобы отрицательные значения не указывали.

rostex replied:
Предполагается, что БД раздает уникальные положительные ID и мне кажется нет особого смыслы в этой валидации

sstani-bgv replied:
У тебя skillDto используется в контролере, как тело запроса. Это тело указывает клиент

DmShevchuk replied:
Валидация не нужна. Потому что id должно генерироваться на стороне сервера, а не клиента. При отправке запроса на создание, id сущности не указывается, передаётся как null

sstani-bgv replied:
> Валидация не нужна. Потому что id должно генерироваться на стороне сервера, а не клиента. При отправке запроса на создание, id сущности не указывается, передаётся как null

Верно, но SkillDto указан как тело запроса при создании скила, и там можно передать id. Нам для создания скила требуется только title => надо создать отдельную dto только с полем title

DmShevchuk replied:
Да, нужно создать разные ДТО на создание и получение сущности.
Так будет гибче, и в кадой ДТО будут использоваться только те поля, которые нужны именно в этом запросе

sstani-bgv commented on src/main/java/school/faang/user_service/service/SkillService.java:
```diff
@@ -0,0 +1,116 @@
+    private final SkillCandidateMapper skillCandidateMapper;
+
+    public SkillDto create(SkillDto skillDto) {
+        if (skillRepository.existsByTitle(skillDto.getTitle())) {
```
По сути у тебя ещё могут совпасть id, я бы их в dto вообще не передавал, пускай бд сама сгенерит

rostex replied:
Не совсем понимаю озвученную проблематику. Подразумевается, что нельзя создать Умение если его название уже существует. Причем тут ID?

sstani-bgv replied:
У тебя передаётся в сигнатуре `SkillDto`, у него 2 поля - id, title => сюда могут передать не null id. Ты далее мапишь dto в `Skill` и т.к. в dto было указано значение id, то и в `Skill` оно будет => при сохранении в бд, если такой id уже существует, то ты перетрёшь старое значение в бд.

rostex replied:
Каким образом? Я проверяю на наличие такого же названия Умения, в БД оно уникальное. при создании сущности мы передаем null ID в БД и уже потом возвращаем с уникальным ID.
Даже если предположить, что каким-то образом мы передадим id в DTO, то при наличии такогоже названия дальше сохранения в БД не будет.

sstani-bgv replied:
Смотри, у тебя этот метод дёргается из контроллера, на контролер запрос будет посылать клиент, он может указать любой id, любой title. Клиент может послать тебе title, которого нет в бд, и id, который есть в бд.

rostex replied:
Проверка условия выполняется только по названию. Наличие ID в данном случае не имеет роли.

DmShevchuk replied:
Проверку на id добавлять не надо

sstani-bgv replied:
> Проверку на id добавлять не надо

Дима, я только что через postman проверил. Я спокойно указал id в теле запроса и всё спокойно создалось. Я оставил такой-же id и поменял title и всё перетёрлось. Тут либо надо проверку на id делать или создать SkillCreateDto, в котором будет одно поле title. И третий вариант - это при мапинге игнорировать поле id

DmShevchuk replied:
Проверку на id делать не надо. После разделения дтошек пропадёт поле id из ДТО на создание сущности.

sstani-bgv commented on src/main/java/school/faang/user_service/service/SkillService.java:
```diff
@@ -0,0 +1,116 @@
+            return null;
+        }
+        List<SkillOffer> skillOffers = skillOfferRepository.findAllOffersOfSkill(skillId, userId);
+        if (skillOffers.size() >= MIN_SKILL_OFFERS) {
```
Лучше сделать if (skillOffers.size() < MIN_SKILL_OFFERS) { // то что под else идёт}, потому что если ты добавишь ещё какие-нибудь if внутрь, то будет много вложенного кода

sstani-bgv commented on src/main/java/school/faang/user_service/service/SkillService.java:
```diff
@@ -0,0 +1,116 @@
+                .orElseThrow(() -> new NoSuchElementException(String.format("Пользователь с ID %d не найден", userId)));
+    }
+
+    private <T> void validateSkillList(List<T> skills) {
```
По моему можно вместо T вставить <?>

rostex replied:
Можно, но в данном конкретном случае считаю избыточным

sstani-bgv replied:
Ты сделаешь более читабельным и кратким код

DmShevchuk replied:
Действительно можно оставить <?>, потому что это упрощает метод и не вводит избыточную обобщенность.

sstani-bgv commented:
У тебя ещё названия коммитов одинаковые)

rostex commented:
> У тебя ещё названия коммитов одинаковые)

Это я в локальную ветку скидывал коммиты. Думал перетрут друг друга, когда в удаленную ветку пушну.

sstani-bgv commented:
> > У тебя ещё названия коммитов одинаковые)
> 
> Это я в локальную ветку скидывал коммиты. Думал перетрут друг друга, когда в удаленную ветку пушну.

Тогда лучше галочку под Amend ставить

rostex committed:
Внесены изменения после кросс-ревью:

Добавлена валидация максимальной длины строки с названием Умения
Откорректирован блок if-else в методе acquireSkillFromOffers
Добавлен тест валидации превышающий максимальную длину строки названия умения

rostex committed:
Внесены изменения после кросс-ревью:

Произведена замена типа элемента в списке в методе валидации

sstani-bgv commented on src/main/java/school/faang/user_service/service/SkillService.java:
```diff
@@ -0,0 +1,116 @@
+            log.info("Недостаточное количество предложений для присвоения умения {}. Необходимо {} вместо {}",
+                    skill.getTitle(), MIN_SKILL_OFFERS, skillOffers.size());
+            return null;
+        } else {
```
Else лучше отпустить

rostex committed:
Внесены изменения после кросс-ревью

rostex committed:
Внесены изменения после кросс-ревью

Добавлено:
- в целях безопасности данных создан SkillCreateDto и SkillCreateMapper с внесением соответствующих изменений в метод create класса SkillService
- внесены соответствущие корректировки в тесты

rostex committed:
Внесены изменения после кросс-ревью

Добавлено:
- в целях безопасности данных создан SkillCreateDto и SkillCreateMapper с внесением соответствующих изменений в метод create класса SkillService
- внесены соответствущие корректировки в тесты

sstani-bgv commented:
А зачем тебе 3 маппера? Оставь только 1 - SkillMapper и добавь туда все методы, просто названия разные сделай и всё

rostex commented:
> А зачем тебе 3 маппера? Оставь только 1 - SkillMapper и добавь туда все методы, просто названия разные сделай и всё

Я художник - я так вижу.

З.Ы. Оптимизировал!

sstani-bgv commented on src/test/java/school/faang/user_service/service/SkillServiceTest.java:
```diff
@@ -0,0 +1,182 @@
+    public void testExistingSkillCreate() {
+        Mockito.when(skillRepository.existsByTitle("Java")).thenReturn(true);
+
+        Assert.assertThrows(
```
`nit` Можно сделать static import, чтобы было не Assert.assertThrows, а assertThrows

rostex replied:
А можно и не делать

sstani-bgv replied:
Просто будет меньше кода)

sstani-bgv commented on src/test/java/school/faang/user_service/service/SkillServiceTest.java:
```diff
@@ -0,0 +1,182 @@
+    private UserRepository userRepository;
+
+    @Mock
+    private SkillMapper skillMapper;
```
Лучше добавить аннотацию Spy, тогда mapper будет реально выполнять свою работу, а не будет заглушкой
```suggestion
    @Spy
    private SkillMapperImpl skillMapper;
```

rostex replied:
Всему свое время

sstani-bgv commented on src/test/java/school/faang/user_service/service/SkillServiceTest.java:
```diff
@@ -0,0 +1,182 @@
+    }
+
+    @Test
+    public void testExistingSkillCreate() {
```
```suggestion
    public void testCreateWithExistingTitle() {
```

sstani-bgv commented on src/test/java/school/faang/user_service/service/SkillServiceTest.java:
```diff
@@ -0,0 +1,182 @@
+
+    @Test
+    public void testNotExistingSkillCreate() {
+        Mockito.when(skillRepository.existsByTitle("Java")).thenReturn(false);
```
Здесь тоже можно сделать static import

rostex replied:
Здесь тоже можно не сделать

rostex committed:
Merge remote-tracking branch 'origin/BJS2-53956/add-skill-operations' into BJS2-53956/add-skill-operations

# Conflicts:
#	src/main/java/school/faang/user_service/mapper/SkillMapper.java
#	src/main/java/school/faang/user_service/service/SkillService.java
#	src/test/java/school/faang/user_service/service/SkillServiceTest.java

rostex committed:
Merge remote-tracking branch 'origin/BJS2-53956/add-skill-operations' into BJS2-53956/add-skill-operations

# Conflicts:
#	src/main/java/school/faang/user_service/mapper/SkillMapper.java
#	src/main/java/school/faang/user_service/service/SkillService.java
#	src/test/java/school/faang/user_service/service/SkillServiceTest.java

rostex committed:
Merge remote-tracking branch 'origin/BJS2-53956/add-skill-operations' into BJS2-53956/add-skill-operations

sstani-bgv commented on src/main/java/school/faang/user_service/service/SkillService.java:
```diff
@@ -0,0 +1,119 @@
+        }
+        Skill skill = skillCreateMapper.toEntity(skillCreateDto);
+        skill = skillRepository.save(skill);
+        return skillMapper.toDto(skill);
```
```suggestion
        return skillMapper.toDto(skillRepository.save(skill));
```

sstani-bgv commented on src/test/java/school/faang/user_service/service/SkillServiceTest.java:
```diff
@@ -0,0 +1,182 @@
+    }
+
+    @Test
+    public void testNotExistingSkillCreate() {
```
```suggestion
    public void testCreateSuccessCase() {
```

sstani-bgv commented on src/main/java/school/faang/user_service/mapper/SkillMapper.java:
`nit` Думаю слово skill в toSkillDto и toSkillCandidateDto можно убрать, т.к. эти методы и так находятся в skillMapper

sstani-bgv commented on src/test/java/school/faang/user_service/service/SkillServiceTest.java:
```diff
@@ -0,0 +1,174 @@
+
+    @Test
+    public void testNotExistingSkillCreate() {
+        Mockito.when(skillRepository.existsByTitle("Java")).thenReturn(false);
+
+        Mockito.when(skillMapper.toEntity(skillCreateDto)).thenReturn(new Skill());
+        skillService.create(skillCreateDto);
+        Skill skill = skillMapper.toEntity(skillCreateDto);
+        Mockito.verify(skillRepository, Mockito.times(1)).save(skill);
```
```suggestion
        Mockito.when(skillRepository.existsByTitle("Java")).thenReturn(false);

        SkillDto dto = skillService.create(skillCreateDto);
        ArgumentCaptor<Skill> argumentCaptor = ArgumentCaptor.forClass(Skill.class);
        Mockito.verify(skillRepository, Mockito.times(1)).save(argumentCaptor.capture());
        Skill capturedSkill = argumentCaptor.getValue();
        
        assertEquals(skillCreateDto.getTitle(), capturedSkill.getTitle());
```
Про ArgumentCaptor есть в лекции про тесты

rostex replied:
Это ошибка?

sstani-bgv replied:
Это более правильная проверка, которая была в лекции

sstani-bgv commented on src/main/java/school/faang/user_service/service/SkillService.java:
```diff
@@ -0,0 +1,115 @@
+
+    public List<SkillDto> getUserSkills(long userId) {
+        List<Skill> skills = skillRepository.findAllByUserId(userId);
+        validateSkillList(skills);
```
По моему это лучше убрать, если у пользователя не будет скилов, то будет просто пустой список

rostex replied:
А по другому считаю

sstani-bgv replied:
Ну просто я считаю, что это проверка ненужная. Когда человек увидит, что ему пришёл пустой список он и так поймёт, что у него нет скилов.  Просто сделаем чуть меньше кода)

sstani-bgv commented on src/test/java/school/faang/user_service/service/SkillServiceTest.java:
```diff
@@ -0,0 +1,174 @@
+    }
+
+    @Test
+    public void testGetUserSkills() {
```
```suggestion
    public void testGetUserSkillsSuccessCase() {
```

sstani-bgv commented on src/main/java/school/faang/user_service/service/SkillService.java:
```diff
@@ -0,0 +1,115 @@
+
+    public List<SkillCandidateDto> getOfferedSkills(long userId) {
+        List<Skill> offeredSkills = skillRepository.findSkillsOfferedToUser(userId);
+        validateSkillList(offeredSkills);
```
Аналогично

rostex replied:
Я вижу 3 строчки кода и слово Аналогично. Очень информативно

sstani-bgv replied:
Прости, тут имелось ввиду тоже убрать проверку на пустой список

sstani-bgv commented on src/main/java/school/faang/user_service/service/SkillService.java:
```diff
@@ -0,0 +1,115 @@
+        List<SkillOffer> skillOffers = skillOfferRepository.findAllOffersOfSkill(skillId, userId);
+        if (skillOffers.size() < MIN_SKILL_OFFERS) {
+            log.info("Недостаточное количество предложений для присвоения умения {}. Необходимо {} вместо {}",
+                    skill.getTitle(), MIN_SKILL_OFFERS, skillOffers.size());
```
По моему здесь надо выкидывать ошибки, где в теле указать причину, чтобы на фронте знали, что произошло и из-за чего. А так просто null возвращается и ничего не известно(в обоих if выкидывать)

rostex replied:
А по мнению авторов задачи Исключения в данном случае не просят выкинуть

sstani-bgv replied:
Ну нам говорили, что не всегда стоит строго следовать условию задачи) Скажи пожалуйста, чем твой вариант лучше моего

sstani-bgv commented on src/test/java/school/faang/user_service/service/SkillServiceTest.java:
```diff
@@ -0,0 +1,174 @@
+    }
+
+    @Test
+    public void testGetOfferedSkills() {
```
```suggestion
    public void testGetOfferedSkillsSuccessCase() {
```

sstani-bgv commented on src/test/java/school/faang/user_service/service/SkillServiceTest.java:
Поменяй названия тестов на такие -  test{название метода}{что тестируешь}. Пример - testCreateWithExistingTitle

rostex committed:
Update src/test/java/school/faang/user_service/service/SkillServiceTest.java

Co-authored-by: Stanislav <stasbugaev2006@gmail.com>

rostex committed:
Update src/test/java/school/faang/user_service/service/SkillServiceTest.java

Co-authored-by: Stanislav <stasbugaev2006@gmail.com>

rostex committed:
Update src/test/java/school/faang/user_service/service/SkillServiceTest.java

Co-authored-by: Stanislav <stasbugaev2006@gmail.com>

rostex committed:
Update src/test/java/school/faang/user_service/service/SkillServiceTest.java

Co-authored-by: Stanislav <stasbugaev2006@gmail.com>

rostex committed:
Внесены изменения в тесты

sstani-bgv commented on src/test/java/school/faang/user_service/service/SkillServiceTest.java:
```diff
@@ -0,0 +1,176 @@
+    private UserRepository userRepository;
+
+    @Spy
+    private SkillMapper skillMapper = Mappers.getMapper(SkillMapper.class);
```
```suggestion
    private SkillMapperImpl skillMapper;
```
`nit` Есть вариант короче записать

DmShevchuk commented on src/main/java/school/faang/user_service/dto/skill/SkillDto.java:
```diff
@@ -0,0 +1,15 @@
+
+@Data
+@NoArgsConstructor
+public class SkillDto {
```
Эта ДТО используется только для чтения.
Нужно переименовать в SkillReadDto

DmShevchuk commented on src/main/java/school/faang/user_service/dto/skill/SkillDto.java:
```diff
@@ -0,0 +1,15 @@
+@NoArgsConstructor
+public class SkillDto {
+    private Long id;
+    @NotBlank(message = "Название умения не может быть пустым")
+    @Size(max = 64, message = "Название умения должно быть не более 64 символов")
```
Валидация в ДТО для чтения не нужна

DmShevchuk commented on src/main/java/school/faang/user_service/execption/DataValidationException.java:
```diff
@@ -0,0 +1,7 @@
+package school.faang.user_service.execption;
+
```
Добавь java-doc комментарий о том, что это за исключение и в каких случаях используется

DmShevchuk commented on src/main/java/school/faang/user_service/service/SkillService.java:
```diff
@@ -0,0 +1,115 @@
+
+    public SkillDto create(SkillCreateDto skillCreateDto) {
+        if (skillRepository.existsByTitle(skillCreateDto.getTitle())) {
+            throw new DataValidationException("Умение с таким названием уже существует");
```
Это не DataValidationException.
DataValidationException подразумевает, что при выполнении кода не прошла валидация. Например, передана пустая строка или число меньше нуля для какого-то поля.

Здесь лучше использовать BusinessException (создать его, если ещё нет), потому что у нас по бизнес-логике не может быть скилов с одинаковым наименованием.

DmShevchuk commented on src/main/java/school/faang/user_service/service/SkillService.java:
```diff
@@ -0,0 +1,115 @@
+
+    public List<SkillDto> getUserSkills(long userId) {
+        List<Skill> skills = skillRepository.findAllByUserId(userId);
+        validateSkillList(skills);
```
Этот метод здесь не нужен. Представь, ты заходишь в корзину на Ozon, в которой нет покупок, и тебе прилетает ошибка во весь экран.

Нет у пользователя скилов, ну и ладно, вернём пустой список.

DmShevchuk commented on src/main/java/school/faang/user_service/service/SkillService.java:
```diff
@@ -0,0 +1,115 @@
+    public List<SkillDto> getUserSkills(long userId) {
+        List<Skill> skills = skillRepository.findAllByUserId(userId);
+        validateSkillList(skills);
+        return skills.stream()
```
Соответственно можно метод в одну строчку написать
```
return skillRepository.findAllByUserId(userId)
          .stream()
          .map(skillMapper::toSkillDto)
          .toList();
```

DmShevchuk commented on src/main/java/school/faang/user_service/service/SkillService.java:
```diff
@@ -0,0 +1,115 @@
+
+    public List<SkillCandidateDto> getOfferedSkills(long userId) {
+        List<Skill> offeredSkills = skillRepository.findSkillsOfferedToUser(userId);
+        validateSkillList(offeredSkills);
```
Здесь тоже проверка на пустой список не нужна

DmShevchuk commented on src/main/java/school/faang/user_service/service/SkillService.java:
```diff
@@ -0,0 +1,115 @@
+        }
+    }
+
+    public static int getMinSkillOffers() {
+        return MIN_SKILL_OFFERS;
+    }
```
Вместо этого попробуй навесить @Getter над полем

DmShevchuk commented on src/main/java/school/faang/user_service/service/SkillService.java:
```diff
@@ -0,0 +1,115 @@
+        if (existingSkill.isPresent()) {
+            log.info("Присвоение умения отклонено, так как умение {} уже существует у пользователя {}",
+                    skill.getTitle(), user.getUsername());
+            return null;
```
Вместо null нужно выбросить BusinessException с соответствующим сообщением об ошибке

DmShevchuk commented on src/main/java/school/faang/user_service/service/SkillService.java:
```diff
@@ -0,0 +1,115 @@
+        if (skillOffers.size() < MIN_SKILL_OFFERS) {
+            log.info("Недостаточное количество предложений для присвоения умения {}. Необходимо {} вместо {}",
+                    skill.getTitle(), MIN_SKILL_OFFERS, skillOffers.size());
+            return null;
```
Вместо null нужно выбросить BusinessException с соответствующим сообщением об ошибке

DmShevchuk commented on src/main/java/school/faang/user_service/service/SkillService.java:
```diff
@@ -0,0 +1,115 @@
+        User user = getUserById(userId);
+        Optional<Skill> existingSkill = skillRepository.findUserSkill(skillId, userId);
+        if (existingSkill.isPresent()) {
+            log.info("Присвоение умения отклонено, так как умение {} уже существует у пользователя {}",
```
log.error

DmShevchuk commented on src/main/java/school/faang/user_service/service/SkillService.java:
```diff
@@ -0,0 +1,115 @@
+        }
+        List<SkillOffer> skillOffers = skillOfferRepository.findAllOffersOfSkill(skillId, userId);
+        if (skillOffers.size() < MIN_SKILL_OFFERS) {
+            log.info("Недостаточное количество предложений для присвоения умения {}. Необходимо {} вместо {}",
```
log.error

rostex committed:
Внесены изменения после ревью

Переименован класс SkillDto в SkillReadDto
Удалена валидация на списки Умений
Добавлен BusinessException
Добавлен JavaDoc в DataValidationException
Откорректированы тесты
